<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>deconvolution.knit</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">VoltRon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="tutorials.html">Explore</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignette
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Data Integration</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="registration.html">Spatial Data Alignment</a>
        </li>
        <li>
          <a href="deconvolution.html">Niche Clustering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Downstream Analysis and Utilities</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="roianalysis.html">ROI Analysis</a>
        </li>
        <li>
          <a href="spotanalysis.html">Cell/Spot Analysis</a>
        </li>
        <li class="dropdown-header">Converting VoltRon Objects</li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-envelope-o"></span>
     
    Contact
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://bioinformatics.mdc-berlin.de">Altuna Lab/BIMSB Bioinfo</a>
    </li>
    <li>
      <a href="https://www.mdc-berlin.de/landthaler">Landthaler Lab/BIMSB</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-github"></span>
     
    GitHub
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://github.com/BIMSBbioinfo/VoltRon">VoltRon</a>
    </li>
    <li>
      <a href="https://github.com/BIMSBbioinfo">BIMSB Bioinfo</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<style>
body {
  text-align: justify
}
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style type="text/css">
.watch-out {
  color: black;
}
</style>
<p><br></p>
<div id="spot-deconvolution-and-niche-clustering"
class="section level1">
<h1>Spot Deconvolution and Niche Clustering</h1>
<p>Essentially, Spots in transcriptomics assays such as <a
href="https://www.10xgenomics.com/products/spatial-gene-expression">Visium</a>
include often some arbitrary number of cells that are of multiple types.
RNA Deconvolution is then incorporated to estimate the
percentage/abundance of these cell types within each spot/ROI given a
reference scRNAseq dataset.</p>
<p>VoltRon includes wrapper commands for using popular RNA deconvolution
methods such as <a
href="https://www.nature.com/articles/s41587-021-00830-w">RCTD</a>
(<strong>spot</strong>) and return estimated abundances as additional
assays within each layer. These novel assays then can be further
processed and used for clustering to detect niches across multiple
datatypes.</p>
<!-- VoltRon includes wrapper commands for using popular RNA deconvolution methods such as [RCTD](https://www.nature.com/articles/s41587-021-00830-w) (**spot**), [SPOTlight](https://academic.oup.com/nar/article/49/9/e50/6129341) (**spot**) and [MuSiC](https://www.nature.com/articles/s41467-018-08023-x) (**ROI**), and return estimated abundances as additional assays within each layer. These novel assays then can be further processed and used for clustering to detect niches across multiple datatypes.   -->
<p><br></p>
<div id="import-st-data" class="section level2">
<h2>Import ST Data</h2>
<p>We will now import the samples and merge them into one VoltRon
object</p>
<pre class="r watch-out"><code>library(VoltRon)

# import Visium data
Vis_A1 &lt;- importVisium(&quot;hColon_N_A1/&quot;, sample_name = &quot;Vis_A1&quot;)
Vis_B1 &lt;- ImportVisium(&quot;hColon_T_B1/&quot;, sample_name = &quot;Vis_B1&quot;)

# merge datasets
Vis_CRC &lt;- list(Vis_A1, Vis_B1)
Vis_CRC &lt;- merge(Vis_CRC[[1]], Vis_CRC[-1])
vrSpatialFeaturePlot(Vis_CRC, features = &quot;Count&quot;, crop = TRUE, alpha = 1, ncol = 2)</code></pre>
<pre><code>VoltRon Object 
Vis_A1: 
  Layers: Section1 
Vis_B1: 
  Layers: Section1 
Assays: Visium(Main) </code></pre>
<p><img width="90%" height="90%" src="images/decon_first_plot.png" class="center"></p>
<p><br></p>
</div>
<div id="import-scrna-data" class="section level2">
<h2>Import scRNA data</h2>
<p>We will now import the scRNA data for reference</p>
<pre class="r watch-out"><code>library(Seurat)
scdata &lt;- readRDS(&quot;GSE166555_uhlitz.rds&quot;)
meta.data &lt;- read.table(&quot;GSE166555_meta_data.tsv&quot;, header = T, sep = &quot;\t&quot;, na.strings = &quot;NA&quot;, fill = TRUE)

# normalize RNA assay before Spot deconvolution
scdata &lt;- Seurat::NormalizeData(scdata, assay = &quot;RNA&quot;)

# Visualize
gtype &lt;- DimPlot(scdata, reduction = &quot;umap&quot;, label = T, group.by = &quot;agg_minorcelltype&quot;) + 
  NoLegend()
gsample &lt;- DimPlot(scdata, reduction = &quot;umap&quot;, label = F, group.by = &quot;SampleType&quot;)
gsample | gtype</code></pre>
<p><img width="90%" height="90%" src="images/decon_singlecell.png" class="center"></p>
<p><br></p>
</div>
<div id="spot-deconvolution-with-rctd" class="section level2">
<h2>Spot Deconvolution with RCTD</h2>
<p>In order to integrate the scRNA data and the Visium data sets within
the VoltRon objects, we will use <strong>RCTD</strong> algorithm which
is accessible with the <a
href="https://github.com/dmcable/spacexr">spacexr</a> package. For all
layers with a Visium data, an additional assay within the same layer
with **_decon** postfix will be created.</p>
<p>It is important at this stage to deconvolute the spots with
phenotypically best suitable single cell reference. Hence, we subset and
create aditional Seurat objects for each individual type of Visium
assay.</p>
<pre class="r watch-out"><code># libraries for RCTD algorithm
library(spacexr)

# Deconvolute normal tissues
scdata_subset_normal &lt;- scdata[,scdata$SampleType == &quot;Normal&quot;]
scdata_subset_normal &lt;- scdata_subset_normal[,!scdata_subset_normal$agg_minorcelltype %in% c(&quot;TC&quot;, &quot;CAFs&quot;, &quot;Pericytes&quot;)]
Vis_CRC_normal &lt;- subset(Vis_CRC, samples = &quot;Vis_A1&quot;)
Vis_CRC_normal &lt;- getDeconvolution(Vis_CRC_normal, sc.object = scdata_subset_normal, 
                                   sc.cluster = &quot;agg_minorcelltype&quot;, max_cores = 6)

# Deconvolute tumor tissues
scdata_subset_tumor &lt;- scdata[,scdata$SampleType == &quot;Tumor&quot;]
Vis_CRC_tumor &lt;- subset(Vis_CRC, samples = &quot;Vis_B1&quot;)
Vis_CRC_tumor &lt;- getDeconvolution(Vis_CRC_tumor, sc.object = scdata_subset_tumor, 
                                  sc.cluster = &quot;agg_minorcelltype&quot;, max_cores = 6)</code></pre>
<p>We now merge two batches of Visium data sets that were deconvoluted
separately.</p>
<pre class="r watch-out"><code># merge samples
Vis_CRC_list &lt;- list(Vis_CRC_normal, Vis_CRC_tumor)
Vis_CRC &lt;- merge(Vis_CRC_list[[1]], Vis_CRC_list[-1])
Vis_CRC</code></pre>
<pre><code>VoltRon Object 
Vis_A1: 
  Layers: Section1 
Vis_B1: 
  Layers: Section1 
Assays: Visium(Main) Visium_decon</code></pre>
<p>We can now switch to the <strong>Visium_decon</strong> assays where
features are now cell types from the scRNA reference and the data values
are cell types percentages in each spot.</p>
<pre class="r watch-out"><code>vrMainAssay(Vis_CRC) &lt;- &quot;Visium_decon&quot;
vrFeatures(Vis_CRC)</code></pre>
<pre><code> [1] &quot;B_cells&quot;     &quot;CB_FBs&quot;      &quot;CCL_FBs&quot;     &quot;CD8_T&quot;       &quot;Endothelial&quot; &quot;Ent_Prog&quot;   
 [7] &quot;Enterocytes&quot; &quot;GC_B_cells&quot;  &quot;Glia&quot;        &quot;Goblet&quot;      &quot;Imma_Gob&quot;    &quot;Mast&quot;       
[13] &quot;Monocyte&quot;    &quot;MyoFBs&quot;      &quot;Plasma_K&quot;    &quot;Plasma_L&quot;    &quot;Stem&quot;        &quot;StemTA&quot;     
[19] &quot;Tconv&quot;       &quot;Treg&quot;        &quot;Tuft&quot;        &quot;UC_FBs&quot;      &quot;CAFs&quot;        &quot;Pericytes&quot;  
[25] &quot;TC&quot;         </code></pre>
<p>These features (i.e. cell type abundances) can be visualized like any
other feature.</p>
<pre class="r watch-out"><code>vrSpatialFeaturePlot(Vis_CRC, features = &quot;MyoFBs&quot;, crop = TRUE, ncol = 5, alpha = 1)</code></pre>
<p><img width="90%" height="90%" src="images/decon_spatialfeature_plot.png" class="center"></p>
<p><br></p>
</div>
<div id="niche-clustering" class="section level2">
<h2>Niche Clustering</h2>
<p>Relative cell type abundances that are learned by RCTD and stored
within VoltRon can now be used to cluster spots. These groups or
clusters of spots now can be realized as <strong>niches</strong>,
regions within tissue sections that have a distinct cell type
composition compared to other groups.</p>
<p>We will now normalize and process relative cell type abundances to
learn these niche clusters. Cell type abundance can be realized as <a
href="https://en.wikipedia.org/wiki/Compositional_data">compositional
data</a>, hence we incorporate <strong>centred log ratio (CLR)</strong>
transformation for normalizing these relative abundances.</p>
<pre class="r watch-out"><code>Vis_CRC &lt;- normalizeData(Vis_CRC, method = &quot;CLR&quot;)</code></pre>
<p>The CLR normalized assay have only 25 features, each representing a
cell type from the single cell reference data. Hence, we can
<strong>directly calculate UMAP reductions from this feature
set</strong> without relying on a prior dimensionality reduction such as
PCA.</p>
<p>VoltRon is also capable of calculating the UMAP reduction from
normalized data slots. Hence, we build a UMAP reduction from CLR data.
However, UMAP will always be calculated from a PCA reduction by default
(if a PCA embedding is found in the object).</p>
<pre class="r watch-out"><code>Vis_CRC &lt;- getUMAP(Vis_CRC, assay = &quot;Visium_decon&quot;, data.type = &quot;norm&quot;)
vrEmbeddingPlot(Vis_CRC, embedding = &quot;umap&quot;, group.by = &quot;Sample&quot;)</code></pre>
<p><img width="60%" height="60%" src="images/decon_embedding_sample.png" class="center"></p>
<p><br></p>
<p>Using normalized cell type abundances, we can now generate k-nearest
neighbor graphs and cluster the graph using leiden method.</p>
<pre class="r watch-out"><code># get neighbors
Vis_CRC &lt;- getNeighbors(Vis_CRC, data.type = &quot;norm&quot;, assay = &quot;Visium_decon&quot;)

# get clustering
Vis_CRC &lt;- getClusters(Vis_CRC, resolution = 0.6, label = &quot;Decon_clusters&quot;)</code></pre>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>VoltRon incorporates distinct plotting functions for,
e.g. embeddings, coordinates, heatmap and even barplots. We can now map
the clusters we have generated on the UMAP embeddings.</p>
<pre class="r watch-out"><code># visualize 
g1 &lt;- vrEmbeddingPlot(Vis_CRC, embedding = &quot;umap&quot;, group.by = &quot;Sample&quot;)
g2 &lt;- vrEmbeddingPlot(Vis_CRC, embedding = &quot;umap&quot;, group.by = &quot;clusters&quot;)
g1 | g2</code></pre>
<p><img width="100%" height="100%" src="images/decon_embedding_clusters.png" class="center"></p>
<p><br></p>
<p>We use <strong>vrHeatmapPlot</strong> to investigate relative cell
type abundances across clusters. We can map these clusters on the tissue
and investigate the locations of spots that are of similar niches.</p>
<pre class="r watch-out"><code>vrSpatialPlot(Vis_CRC, group.by = &quot;clusters&quot;, crop = TRUE, alpha = 1)
vrHeatmapPlot(Vis_CRC, features = vrFeatures(Vis_CRC_merged), group.by = &quot;clusters&quot;, 
              show_row_names = T, show_heatmap_legend = T)</code></pre>
<!-- <table> -->
<!-- <tr> -->
<!--   <td width="53%"><img width="100%" height="100%" src="images/decon_spatial_clusters.png" class="center"> </td> -->
<!--   <td width="46%"><img width="100%" height="100%" src="images/decon_heatmap_clusters.png" class="center"> </td> -->
<!-- </tr> -->
<!-- </table> -->
<p><img width="80%" height="80%" src="images/decon_spatial_clusters.png" class="center">
<img width="80%" height="80%" src="images/decon_heatmap_clusters.png" class="center">
<br></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
