---
title: "Image Registration"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

<style>
.title{
  display: none;
}
body {
  text-align: justify
}
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{css, echo=FALSE}
.watch-out {
  color: black;
}
```

```{r setup, include=FALSE}
# use rmarkdown::render_site(envir = knitr::knit_global())
knitr::opts_chunk$set(highlight = TRUE, echo = TRUE)
```

<br>

# Spatial Data Alignment

Spatial genomic technologies often generate diverse images and spatial readouts, even though the tissue slices are from adjacent sections of a single tissue block. Hence, the alignment of images and spatial coordinates across tissue sections are of utmost importance to dissect the correct spatial closeness across 


Hence, any spatial data analysis across entities (ROIs, spots, cells, molecules etc.) would be inconclusive due to 

VoltRon allows users to **align spatial omics datasets of these serial sections** for data transfer and 3 dimensional stack alignment. The order of the tissue/sample slices should be provided by the user. VoltRon provides a fully embedded **shiny application** to either automatically or manually align images. The automatic alignment is achieved by the **OpenCV**'s C++ library fully embedded in the VoltRon package.

<table>
<tbody>
  <tr style = "vertical-align: center">
  <td style = "width:43%; vertical-align: center"> <img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/manualregistration.png" class="center"></td>
  <td style = "width:43%; vertical-align: center"> <img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/autoregistration.png" class="center"></td>
  </tr>
</tbody>
</table>

<br>

## Alignment of Xenium and Visium 

In this use case, we will align **immunofluorescence (IF)** and **H&E images** of the **Xenium In Situ and Visium CytAssist platforms** readouts. Three tissue sections are derived from a single formalin-fixed, paraffin-embedded (FFPE) breast cancer tissue block. A 5 $\mu$m section was taken for Visium CytAssist and two replicate 5 $\mu$m sections were taken for the Xenium replicates. 

Xenium and Visium readouts can be downloaded from [here](https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast) (specifically, import **In Situ Replicate 1/2 and Visium Spatial**). Alternatively, you can **download a zipped collection of three Visium and Xenium readouts** from [here](https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/SpatialDataAlignment/Xenium_vs_Visium/10X_Xenium_Visium.zip). More information on the spatial datasets and the study can be also be found on the [BioArxiv preprint](https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1).

VoltRon includes built-in functions to converting single readouts from Xenium and Visium platforms into VoltRon objects. We will import both Xenium replicates alongside with the Visium CytAssist data so that we can register images of these assays and merge them into one VoltRon object. 

```{r eval = FALSE, class.source="watch-out"}
library(VoltRon)
Xen_seu_R1 <- importXenium("XeniumR1/outs", sample_name = "XeniumR1")
Xen_seu_R2 <- importXenium("XeniumR2/outs", sample_name = "XeniumR2")
Vis_seu <- importVisium("VisiumR1/outs", sample_name = "VisiumR1")
```

```{r eval = FALSE, include=FALSE, class.source="watch-out"}
library(Seurat)

# Load Xenium Data
Xen_seu_R1 <- importXenium("../../../data/10X_Xenium_Visium/Xenium_R1/outs", sample_name = "XeniumR1")

# Load Xenium Data
Xen_seu_R2 <- importXenium("../../../data/10X_Xenium_Visium/Xenium_R2/outs", sample_name = "XeniumR2")

# Load Visium Data
Vis_seu <- importVisium("../../../data/10X_Xenium_Visium/Visium", sample_name = "VisiumR1")

# saved keypoints
keypoints <- readRDS("../../../workflows/saveddata/Xenium&Visium_keypoints.rds")
keypoints <- list(`1-2` = list(ref = keypoints[[2]], query = keypoints[[1]]),
                  `2-3` = list(ref = keypoints[[1]], query = keypoints[[3]]))
xen_reg <- registerSpatialData(data_list = list(Xen_seu_R1, Vis_seu, Xen_seu_R2), keypoints = keypoints)
```

<br>

### Manual Image Alignment

In order to achieve data transfer and integration across these two modalities, we need to first make sure that spatial coordinate spaces of these three datasets are aligned. For this, we will make use of the **registerSpatialData** function which calls a small **shiny app** embedded into VoltRon. The function takes a single list as an input where the order of VoltRon objects in the list should be the same as the **order of serial sections**. 

In this example, we will **manually choose keypoints (or landmarks)** that are observed to be common across pairs of images. The image in the middle (Visium CytAssist) would be taken as the image of reference, and hence all other images (or spatial datasets) are to be aligned to the Visium data. **registerSpatialData** will return a list of VoltRon objects whose assays include both the original and registered versions of spatial coordinates and images.

```{r class.source="watch-out", eval = FALSE} 
xen_reg <- registerSpatialData(data_list = list(Xen_seu_R1, Vis_seu, Xen_seu_R2))
```

<img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/XeniumVisiumRegistration.gif" class="center">

You can use the same keypoints you have selected in the first run, and reproduce the manual alignment without choosing for the second time. 

```{r class.source="watch-out", eval = FALSE} 
keypoints <- xen_reg$keypoints
xen_reg <- registerSpatialData(data_list = list(Xen_seu_R1, Vis_seu, Xen_seu_R2), keypoints = keypoints)
```

```{r class.source="watch-out", eval = FALSE, include=FALSE}
keypoints <- readRDS("saved_keypoints/Xenium&Visium_keypoints.rds")
keypoints <- list(`1-2` = list(ref = keypoints[[2]], query = keypoints[[1]]),
                  `2-3` = list(ref = keypoints[[1]], query = keypoints[[3]]))
xen_reg <- registerSpatialData(data_list = list(Xen_seu_R1, Vis_seu, Xen_seu_R2), keypoints = keypoints)
```

<br>

### Combine VoltRon object

Now that the VoltRon objects of Xenium and Visium datasets are accurately aligned, we can combine these objects to create **one VoltRon object with three layers**. Since all sections are derived from the same tissue block, we want them to be associated with the same sample, hence we define the sample name as well. VoltRon will recognize that all layers are originated from the same sample/block, and choose the majority assay as the main assay. 

```{r class.source="watch-out", eval = FALSE}
merge_list <- xen_reg$registered_spat
SRBlock <- merge(merge_list[[1]], merge_list[-1], sample_name = "10XBlock")
SRBlock
```
```
10XBlock: 
  Layers: Section1 Section2 Section3 
Assays: Xenium (Main) Visium
```

<br>

### Data/Label Transfer

The combined VoltRon object of Visium and Xenium datasets can be used to transfer information across layers and assays. This is accomplished by aggregating and summarizing, for example, gene counts of cells from the Xenium assay aligned with each spots of Visium assay. 

Either labels or cell types can be summarized to generate; 

* pseudo cell type abundance assays or
* pseudo gene expression assays 

given the same spot orientation of the adjacent Visium layer **(under development)**.

```{r class.source="watch-out", eval = FALSE}
```

<br>

## Alignment of DLPFC Visium

In the next use case, we will align H&E images associated with Visium data generated from tissue block sections of adult humans with postmortem dorsolateral prefrontal cortex (DLPFC). Two pairs of adjacent sections, that are 10 $\mu$m away from each other, was obtained from the tissue block of the third donor. Hence, we align each pair individually. The datasets can be download from [here](https://research.libd.org/spatialLIBD/). 

```{r class.source="watch-out", eval = FALSE} 
library(VoltRon)
DLPFC_1 <- importVisium("DLPFC/151673", sample_name = "DLPFC_1")
DLPFC_2 <- importVisium("DLPFC/151674", sample_name = "DLPFC_2")
DLPFC_3 <- importVisium("DLPFC/151675", sample_name = "DLPFC_3")
DLPFC_4 <- importVisium("DLPFC/151676", sample_name = "DLPFC_4")
```

<br>

### Automated Image Alignment

We will use the registerSpatialData function to **automatically register two Visium assays (two H&E images)** without choosing the landmark points. The shiny app will provide **two images** for this task: 

* An image that shows the automated matched points across two images, and 
* A slideshow with both registered images that demonstrates the accuracy of the alignment. 

This app also provide two tuning parameters that used by the **OpenCV's image registration** workflow fully embedded in VoltRon: 

* **# of Features** option specifies the number of maximum image features spotted within each image which later be used to match these across multiple images.
* **Match %** specifies the percentage of these features matching at max which in turn used to compute the registration/transformation matrix. 

The registration matrix is used to convert the spatial coordinates of the VoltRon object to align coordinates to a single reference. 

We will use **1000 features** for this alignment, set **Match %** to \%20 of the features to be matched across images. The quality of the alignment will be determined by the fine tuning of these parameters where users will immediately observe the alignment quality by sliding registered images. 

```{r class.source="watch-out", eval = FALSE} 
DLPFC_list <- list(DLPFC_1, DLPFC_2)
reg1and2 <- registerSpatialData(data_list = DLPFC_list)
```

<img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/VisiumDLFPCRegistration.gif" class="center">

<br>

We can now apply a similar alignment across the second pair of VoltRon objects. We will use 800 features for this alignment,  set "Match %" to \%50 of the features to be matched across images.

```{r class.source="watch-out", eval = FALSE} 
DLPFC_list <- list(DLPFC_3, DLPFC_4)
reg3and4 <- registerSpatialData(data_list = DLPFC_list)
```

<br>

### Combine VoltRon object

There are two pairs of adjacent sections which we aligned each pairs separately. We can combine these two lists into one list and merge VoltRon objects. 

```{r class.source="watch-out", eval = FALSE} 
merge_list <- c(reg1and2$registered_spat, reg3and4$registered_spat)
SRBlock <- merge(merge_list[[1]], merge_list[-1], sample_name = "DLPFC_Block")
SRBlock
```

```
VoltRon Object 
10XBlock: 
  Layers: Section1 Section2 
Assays: Visium (Main) 
```

<br>

### 3D Spot Clustering 

Aligning spots along the z-stack allows us to cluster these spots using both the gene expression similarities and spatial adjacency (both along the x-y direction and in the z direction). We first generate a spatial neighborhood graph and use this graph along with the gene expression neighborhood graph **(under Development)**.




