<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Image Registration</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">VoltRon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="tutorials.html">Explore</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignette
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Data Integration</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="registration.html">Spatial Data Alignment</a>
        </li>
        <li>
          <a href="deconvolution.html">Niche Clustering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Downstream Analysis and Utilities</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="roianalysis.html">ROI Analysis</a>
        </li>
        <li>
          <a href="spotanalysis.html">Cell/Spot Analysis</a>
        </li>
        <li class="dropdown-header">Converting VoltRon Objects</li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-envelope-o"></span>
     
    Contact
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://bioinformatics.mdc-berlin.de">Altuna Lab/BIMSB Bioinfo</a>
    </li>
    <li>
      <a href="https://www.mdc-berlin.de/landthaler">Landthaler Lab/BIMSB</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-github"></span>
     
    GitHub
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://github.com/BIMSBbioinfo/VoltRon">VoltRon</a>
    </li>
    <li>
      <a href="https://github.com/BIMSBbioinfo">BIMSB Bioinfo</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Image Registration</h1>

</div>


<style>
.title{
  display: none;
}
body {
  text-align: justify
}
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style type="text/css">
.watch-out {
  color: black;
}
</style>
<p><br></p>
<div id="spatial-data-alignment" class="section level1">
<h1>Spatial Data Alignment</h1>
<p>Spatial genomic technologies often generate diverse images and
spatial readouts, even though the tissue slices are from adjacent
sections of a single tissue block. Hence, the alignment of images and
spatial coordinates across tissue sections are of utmost importance to
dissect the correct spatial closeness across</p>
<p>Hence, any spatial data analysis across entities (ROIs, spots, cells,
molecules etc.) would be inconclusive due to</p>
<p>VoltRon allows users to <strong>align spatial omics datasets of these
serial sections</strong> for data transfer and 3 dimensional stack
alignment. The order of the tissue/sample slices should be provided by
the user. VoltRon provides a fully embedded <strong>shiny
application</strong> to either automatically or manually align images.
The automatic alignment is achieved by the <strong>OpenCV</strong>’s C++
library fully embedded in the VoltRon package.</p>
<table>
<tbody>
<tr style="vertical-align: center">
<td style="width:43%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/manualregistration.png" class="center">
</td>
<td style="width:43%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/autoregistration.png" class="center">
</td>
</tr>
</tbody>
</table>
<p><br></p>
<div id="alignment-of-xenium-and-visium" class="section level2">
<h2>Alignment of Xenium and Visium</h2>
<p>In this use case, we will align <strong>immunofluorescence
(IF)</strong> and <strong>H&amp;E images</strong> of the <strong>Xenium
In Situ and Visium CytAssist platforms</strong> readouts. Three tissue
sections are derived from a single formalin-fixed, paraffin-embedded
(FFPE) breast cancer tissue block. A 5 <span
class="math inline">\(\mu\)</span>m section was taken for Visium
CytAssist and two replicate 5 <span class="math inline">\(\mu\)</span>m
sections were taken for the Xenium replicates.</p>
<p>Xenium and Visium readouts can be downloaded from <a
href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast">here</a>
(specifically, import <strong>In Situ Replicate 1/2 and Visium
Spatial</strong>). Alternatively, you can <strong>download a zipped
collection of three Visium and Xenium readouts</strong> from <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/SpatialDataAlignment/Xenium_vs_Visium/10X_Xenium_Visium.zip">here</a>.
More information on the spatial datasets and the study can be also be
found on the <a
href="https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1">BioArxiv
preprint</a>.</p>
<p>VoltRon includes built-in functions to converting single readouts
from Xenium and Visium platforms into VoltRon objects. We will import
both Xenium replicates alongside with the Visium CytAssist data so that
we can register images of these assays and merge them into one VoltRon
object.</p>
<pre class="r watch-out"><code>library(VoltRon)
Xen_seu_R1 &lt;- importXenium(&quot;XeniumR1/outs&quot;, sample_name = &quot;XeniumR1&quot;)
Xen_seu_R2 &lt;- importXenium(&quot;XeniumR2/outs&quot;, sample_name = &quot;XeniumR2&quot;)
Vis_seu &lt;- importVisium(&quot;VisiumR1/outs&quot;, sample_name = &quot;VisiumR1&quot;)</code></pre>
<p><br></p>
<div id="manual-image-alignment" class="section level3">
<h3>Manual Image Alignment</h3>
<p>In order to achieve data transfer and integration across these two
modalities, we need to first make sure that spatial coordinate spaces of
these three datasets are aligned. For this, we will make use of the
<strong>registerSpatialData</strong> function which calls a small
<strong>shiny app</strong> embedded into VoltRon. The function takes a
single list as an input where the order of VoltRon objects in the list
should be the same as the <strong>order of serial sections</strong>.</p>
<p>In this example, we will <strong>manually choose keypoints (or
landmarks)</strong> that are observed to be common across pairs of
images. The image in the middle (Visium CytAssist) would be taken as the
image of reference, and hence all other images (or spatial datasets) are
to be aligned to the Visium data. <strong>registerSpatialData</strong>
will return a list of VoltRon objects whose assays include both the
original and registered versions of spatial coordinates and images.</p>
<pre class="r watch-out"><code>xen_reg &lt;- registerSpatialData(data_list = list(Xen_seu_R1, Vis_seu, Xen_seu_R2))</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/XeniumVisiumRegistration.gif" class="center"></p>
<p>You can use the same keypoints you have selected in the first run,
and reproduce the manual alignment without choosing for the second
time.</p>
<pre class="r watch-out"><code>keypoints &lt;- xen_reg$keypoints
xen_reg &lt;- registerSpatialData(data_list = list(Xen_seu_R1, Vis_seu, Xen_seu_R2), keypoints = keypoints)</code></pre>
<p><br></p>
</div>
<div id="combine-voltron-object" class="section level3">
<h3>Combine VoltRon object</h3>
<p>Now that the VoltRon objects of Xenium and Visium datasets are
accurately aligned, we can combine these objects to create <strong>one
VoltRon object with three layers</strong>. Since all sections are
derived from the same tissue block, we want them to be associated with
the same sample, hence we define the sample name as well. VoltRon will
recognize that all layers are originated from the same sample/block, and
choose the majority assay as the main assay.</p>
<pre class="r watch-out"><code>merge_list &lt;- xen_reg$registered_spat
SRBlock &lt;- merge(merge_list[[1]], merge_list[-1], sample_name = &quot;10XBlock&quot;)
SRBlock</code></pre>
<pre><code>10XBlock: 
  Layers: Section1 Section2 Section3 
Assays: Xenium (Main) Visium</code></pre>
<p><br></p>
</div>
<div id="datalabel-transfer" class="section level3">
<h3>Data/Label Transfer</h3>
<p>The combined VoltRon object of Visium and Xenium datasets can be used
to transfer information across layers and assays. This is accomplished
by aggregating and summarizing, for example, gene counts of cells from
the Xenium assay aligned with each spots of Visium assay.</p>
<p>Either labels or cell types can be summarized to generate;</p>
<ul>
<li>pseudo cell type abundance assays or</li>
<li>pseudo gene expression assays</li>
</ul>
<p>given the same spot orientation of the adjacent Visium layer
<strong>(under development)</strong>.</p>
<p><br></p>
</div>
</div>
<div id="alignment-of-dlpfc-visium" class="section level2">
<h2>Alignment of DLPFC Visium</h2>
<p>In the next use case, we will align H&amp;E images associated with
Visium data generated from tissue block sections of adult humans with
postmortem dorsolateral prefrontal cortex (DLPFC). Two pairs of adjacent
sections, that are 10 <span class="math inline">\(\mu\)</span>m away
from each other, was obtained from the tissue block of the third donor.
Hence, we align each pair individually. The datasets can be download
from <a href="https://research.libd.org/spatialLIBD/">here</a>.</p>
<pre class="r watch-out"><code>library(VoltRon)
DLPFC_1 &lt;- importVisium(&quot;DLPFC/151673&quot;, sample_name = &quot;DLPFC_1&quot;)
DLPFC_2 &lt;- importVisium(&quot;DLPFC/151674&quot;, sample_name = &quot;DLPFC_2&quot;)
DLPFC_3 &lt;- importVisium(&quot;DLPFC/151675&quot;, sample_name = &quot;DLPFC_3&quot;)
DLPFC_4 &lt;- importVisium(&quot;DLPFC/151676&quot;, sample_name = &quot;DLPFC_4&quot;)</code></pre>
<p><br></p>
<div id="automated-image-alignment" class="section level3">
<h3>Automated Image Alignment</h3>
<p>We will use the registerSpatialData function to <strong>automatically
register two Visium assays (two H&amp;E images)</strong> without
choosing the landmark points. The shiny app will provide <strong>two
images</strong> for this task:</p>
<ul>
<li>An image that shows the automated matched points across two images,
and</li>
<li>A slideshow with both registered images that demonstrates the
accuracy of the alignment.</li>
</ul>
<p>This app also provide two tuning parameters that used by the
<strong>OpenCV’s image registration</strong> workflow fully embedded in
VoltRon:</p>
<ul>
<li><strong># of Features</strong> option specifies the number of
maximum image features spotted within each image which later be used to
match these across multiple images.</li>
<li><strong>Match %</strong> specifies the percentage of these features
matching at max which in turn used to compute the
registration/transformation matrix.</li>
</ul>
<p>The registration matrix is used to convert the spatial coordinates of
the VoltRon object to align coordinates to a single reference.</p>
<p>We will use <strong>1000 features</strong> for this alignment, set
<strong>Match %</strong> to %20 of the features to be matched across
images. The quality of the alignment will be determined by the fine
tuning of these parameters where users will immediately observe the
alignment quality by sliding registered images.</p>
<pre class="r watch-out"><code>DLPFC_list &lt;- list(DLPFC_1, DLPFC_2)
reg1and2 &lt;- registerSpatialData(data_list = DLPFC_list)</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/VisiumDLFPCRegistration.gif" class="center"></p>
<p><br></p>
<p>We can now apply a similar alignment across the second pair of
VoltRon objects. We will use 800 features for this alignment, set “Match
%” to %50 of the features to be matched across images.</p>
<pre class="r watch-out"><code>DLPFC_list &lt;- list(DLPFC_3, DLPFC_4)
reg3and4 &lt;- registerSpatialData(data_list = DLPFC_list)</code></pre>
<p><br></p>
</div>
<div id="combine-voltron-object-1" class="section level3">
<h3>Combine VoltRon object</h3>
<p>There are two pairs of adjacent sections which we aligned each pairs
separately. We can combine these two lists into one list and merge
VoltRon objects.</p>
<pre class="r watch-out"><code>merge_list &lt;- c(reg1and2$registered_spat, reg3and4$registered_spat)
SRBlock &lt;- merge(merge_list[[1]], merge_list[-1], sample_name = &quot;DLPFC_Block&quot;)
SRBlock</code></pre>
<pre><code>VoltRon Object 
10XBlock: 
  Layers: Section1 Section2 
Assays: Visium (Main) </code></pre>
<p><br></p>
</div>
<div id="d-spot-clustering" class="section level3">
<h3>3D Spot Clustering</h3>
<p>Aligning spots along the z-stack allows us to cluster these spots
using both the gene expression similarities and spatial adjacency (both
along the x-y direction and in the z direction). We first generate a
spatial neighborhood graph and use this graph along with the gene
expression neighborhood graph <strong>(under Development)</strong>.</p>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
