<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>registration.knit</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">VoltRon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="tutorials.html">Explore</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignette
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Data Integration</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="registration.html">Spatial Data Alignment</a>
        </li>
        <li>
          <a href="deconvolution.html">Niche Clustering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Downstream Analysis and Utilities</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="spotanalysis.html">Cell/Spot Analysis</a>
        </li>
        <li>
          <a href="roianalysis.html">ROI Analysis</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-envelope-o"></span>
     
    Contact
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://bioinformatics.mdc-berlin.de">Altuna Lab/BIMSB Bioinfo</a>
    </li>
    <li>
      <a href="https://www.mdc-berlin.de/landthaler">Landthaler Lab/BIMSB</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-github"></span>
     
    GitHub
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://github.com/BIMSBbioinfo/VoltRon">VoltRon</a>
    </li>
    <li>
      <a href="https://github.com/BIMSBbioinfo">BIMSBbioinfo</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<style>
body {
  text-align: justify
}
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style type="text/css">
.watch-out {
  color: black;
}
</style>
<p><br></p>
<div id="spatial-data-alignment" class="section level1">
<h1>Spatial Data Alignment</h1>
<p>At its core, VoltRon allows users to align multiple consecutive
spatial data slices for data transfer and 3 dimensional stack alignment.
The order of the tissue/sample slices, and the prior information on the
adjacency of these slices, should be provided by the user. VoltRon
provides a fully embedded shiny application to allow users to either
automatically or manually align consecutive 2d spatial datasets.</p>
<p>Using both automated or manual alignment options, users are given the
option to tune the alignment parameters to check the accuracy of the
alignment.</p>
<p><br></p>
<table>
<tbody>
<tr style="vertical-align: center">
<td style="width:43%; vertical-align: center">
<img src="images/manualregistration.png" class="center">
</td>
<td style="width:43%; vertical-align: center">
<img src="images/autoregistration.png" class="center">
</td>
</tr>
</tbody>
</table>
<p><br></p>
<div id="xenium-and-visium" class="section level2">
<h2>Xenium and Visium</h2>
<p>In this use case, we will align immunofluorescence (IF) and H&amp;E
images associated with the readouts of Xenium In Situ and Visium
CytAssist platforms. Tissue sections are derived from a single
formalin-fixed, paraffin-embedded (FFPE) breast cancer tissue block
where 3 serial sections of 5 <span class="math inline">\(\mu\)</span>m
thickness. A 5 <span class="math inline">\(\mu\)</span>m section was
taken for Visium CytAssist and two replicate 5 <span
class="math inline">\(\mu\)</span>m sections were taken for Xenium. The
dataset is download from <a
href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast">here</a>.</p>
<p>More information on the dataset and the study can be also be found on
the <a
href="https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1">BioArxiv
preprint</a>.</p>
<p><br></p>
<div id="import" class="section level3">
<h3>Import</h3>
<p>VoltRon includes built-in functions to converting readouts from both
Xenium and Visium platforms into VoltRon objects. We will import both
Xenium replicates alongside with the Visium CytAssist data.</p>
<pre class="r watch-out"><code>library(VoltRon)
Xen_seu_R1 &lt;- importXenium(&quot;Xenium_R1/outs&quot;, sample_name = &quot;XeniumR1&quot;)
Xen_seu_R2 &lt;- importXenium(&quot;Xenium_R2/outs&quot;, sample_name = &quot;XeniumR2&quot;)
Vis_seu &lt;- importVisium(&quot;Visium/outs&quot;, sample_name = &quot;VisiumR1&quot;)</code></pre>
<p><br></p>
</div>
<div id="image-alignment" class="section level3">
<h3>Image Alignment</h3>
<p>In order to achieve data transfer and integration across these two
modalities, we need to first make sure that spatial coordinate spaces of
these three datasets are aligned. For this, we will make use of the
<strong>registerSpatialData</strong> function which calls a small shiny
app embedded into VoltRon. Users have to provide the list of VoltRon
objects associated with each dataset where the order of components of
the list should be the same as the serial order of the sections within
the tissue/sample block.</p>
<p>In this example, we will manually choose keypoints (or landmarks)
that are observed to be common across pairs of images. The image in the
middle (Visium CytAssist) would be taken as the image of reference, and
hence all other images (or spatial datasets) are to be aligned to the
Visium data. <strong>registerSpatialData</strong> will return a list of
VoltRon objects whose assays include both the original and registered
versions of spatial coordinates and images.</p>
<pre class="r watch-out"><code>xen_reg &lt;- registerSpatialData(data_list = list(Xen_seu_R1, Vis_seu, Xen_seu_R2))</code></pre>
<p><img width="80%" height="80%" src="images/XeniumVisiumRegistration.gif" class="center"></p>
<p>The returned list will also include a set of keypoints selected by
the user if they have used manual alignment. You can use this keypoints
object again to call <strong>registerSpatialData</strong> and reproduce
the earlier manual alignment.</p>
<pre class="r watch-out"><code>keypoints &lt;- xen_reg$keypoints
xen_reg &lt;- registerSpatialData(data_list = list(Xen_seu_R1, Vis_seu, Xen_seu_R2), keypoints = keypoints)</code></pre>
<p><br></p>
</div>
<div id="combine-voltron-object" class="section level3">
<h3>Combine VoltRon object</h3>
<p>Now that the VoltRon objects of Xenium and Visium datasets are
accurately aligned, we can combine these objects to create one VoltRon
object with three layers. Since all sections are derived from the same
tissue block, we want them to be associated with the same sample, hence
we define the sample name as well.</p>
<p>VoltRon will pick the assay type that is of the majority across all
layers across the VoltRon object and appoint that as the main assay.</p>
<pre class="r watch-out"><code>merge_list &lt;- xen_reg$registered_spat
SRBlock &lt;- merge(merge_list[[1]], merge_list[-1], sample_name = &quot;10XBlock&quot;)
SRBlock</code></pre>
<pre><code>10XBlock: 
  Layers: Section1 Section2 Section3 
Assays: Xenium (Main) Visium</code></pre>
<p><br></p>
</div>
</div>
<div id="alignment-of-dlpfc-visium" class="section level2">
<h2>Alignment of DLPFC Visium</h2>
<p>In the next use case, we will align H&amp;E images associated with
Visium data generated from tissue block sections of adult humans with
postmortem dorsolateral prefrontal cortex (DLPFC). Two pairs of adjacent
sections, that are 10 <span class="math inline">\(\mu\)</span>m away
from each other, was obtained from the tissue block of the third donor.
Hence, we align each pair individually.</p>
<div id="import-1" class="section level3">
<h3>Import</h3>
<pre class="r watch-out"><code>library(VoltRon)
DLPFC_1 &lt;- importVisium(&quot;../../../data/10X_Visium_DLPFC/151673&quot;, sample_name = &quot;DLPFC_1&quot;)
DLPFC_2 &lt;- importVisium(&quot;../../../data/10X_Visium_DLPFC/151674&quot;, sample_name = &quot;DLPFC_2&quot;)
DLPFC_3 &lt;- importVisium(&quot;../../../data/10X_Visium_DLPFC/151675&quot;, sample_name = &quot;DLPFC_3&quot;)
DLPFC_4 &lt;- importVisium(&quot;../../../data/10X_Visium_DLPFC/151676&quot;, sample_name = &quot;DLPFC_4&quot;)</code></pre>
<p><br></p>
</div>
<div id="automated-image-alignment" class="section level3">
<h3>Automated Image Alignment</h3>
<p>We will use the <strong>registerSpatialData</strong> function to
automatically register H&amp;E images, and hence two VoltRon objects, to
each other, without choosing the landmark points. The shiny app will
provide two images for this task: (i) the matching image that shows the
automated matched points across two images, and (ii) a slideshow with
both registered images that demonstrates the accuracy of the
alignment.</p>
<p>This app also provide two tuning parameters that used by the OpenCV’s
image registration workflow fully embedded in VoltRon. The <strong># of
Features</strong> option specifies the number of maximum image features
spotted within each image which later be used to match these across
multiple images, and the <strong>Match %</strong> specifies part
percentage of these features should be matching at max to be used to
compute the registeration/transformation matrix. The same matrix is used
to convert the spatial coordinates of the VoltRon object to align
coordinates to a single reference.</p>
<p>We will use 1000 features for this alignment, ask for %20 of the
features to be matched across images. The quality of the alignment will
be determined by the fine tuning of these parameters where users will
immediately observe the alignment quality by sliding registered
images.</p>
<pre class="r watch-out"><code>DLPFC_list &lt;- list(DLPFC_1, DLPFC_2)
reg1and2 &lt;- registerSpatialData(data_list = DLPFC_list)</code></pre>
<p><img width="80%" height="80%" src="images/VisiumDLFPCRegistration.gif" class="center"></p>
<p><br></p>
<p>We can now apply a similar alignment across the second pair of
VoltRon objects. We will use 800 features for this alignment, ask for
%50 of the features to be matched across images.</p>
<pre class="r watch-out"><code>DLPFC_list &lt;- list(DLPFC_3, DLPFC_4)
reg3and4 &lt;- registerSpatialData(data_list = DLPFC_list)</code></pre>
</div>
<div id="combine-voltron-object-1" class="section level3">
<h3>Combine VoltRon object</h3>
<p>We can now combine these two spatial datasets and make a block of
Visium slices. Since there is only a one type of assay, which is Visium,
that would be the main assay.</p>
<pre class="r watch-out"><code>merge_list &lt;- c(reg1and2$registered_spat, reg3and4$registered_spat)
SRBlock &lt;- merge(merge_list[[1]], merge_list[-1], sample_name = &quot;DLPFC_Block_R3&quot;)
SRBlock</code></pre>
<pre><code>VoltRon Object 
10XBlock: 
  Layers: Section1 Section2 
Assays: Visium (Main) </code></pre>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
