<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Cell/Spot Analysis</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">VoltRon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="tutorials.html">Explore</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignette
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Data Integration</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="registration.html">Spatial Data Alignment</a>
        </li>
        <li>
          <a href="deconvolution.html">Niche Clustering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Downstream Analysis and Utilities</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="roianalysis.html">ROI Analysis</a>
        </li>
        <li>
          <a href="spotanalysis.html">Cell/Spot Analysis</a>
        </li>
        <li class="dropdown-header">Converting VoltRon Objects</li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-envelope-o"></span>
     
    Contact
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://bioinformatics.mdc-berlin.de">Altuna Lab/BIMSB Bioinfo</a>
    </li>
    <li>
      <a href="https://www.mdc-berlin.de/landthaler">Landthaler Lab/BIMSB</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-github"></span>
     
    GitHub
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://github.com/BIMSBbioinfo/VoltRon">VoltRon</a>
    </li>
    <li>
      <a href="https://github.com/BIMSBbioinfo">BIMSB Bioinfo</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Cell/Spot Analysis</h1>

</div>


<style>
.title{
  display: none;
}
body {
  text-align: justify
}
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style type="text/css">
.watch-out {
  color: black;
}
</style>
<p><br></p>
<div id="xenium-data-analysis" class="section level1">
<h1>Xenium Data Analysis</h1>
<p>VoltRon is an end-to-end spatial data analysis package which also
supports investigating spatial points in single cell resolution. VoltRon
includes essential built-in functions capable of
<strong>filtering</strong>, <strong>processing</strong> and
<strong>clustering</strong> as well as <strong>visualizing</strong>
spatial datasets with a goal of cell type discovery and annotation.</p>
<p>In this use case, we analyse readouts of the <strong>Xenium In
Situ</strong> platform. Two tissue sections are derived from a single
formalin-fixed, paraffin-embedded (FFPE) breast cancer tissue block. Two
replicate of 5 <span class="math inline">\(\mu\)</span>m sections were
taken for the Xenium replicates. More information on the spatial
datasets and the study can be also be found on the <a
href="https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1">BioArxiv
preprint</a>.</p>
<p>You can import these readouts from the <a
href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast">10x
Genomics website</a> (specifically, import <strong>In Situ Replicate
1/2</strong>). Alternatively, you can <strong>download a zipped
collection of Xenium readouts</strong> from <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/SpatialDataAlignment/Xenium_vs_Visium/10X_Xenium_Visium.zip">here</a>.</p>
<div id="building-voltron-objects" class="section level2">
<h2>Building VoltRon objects</h2>
<p>VoltRon includes built-in functions for converting readouts from both
Xenium into VoltRon objects.</p>
<pre class="r watch-out"><code>library(VoltRon)
Xen_R1 &lt;- importXenium(&quot;Xenium_R1/outs&quot;, sample_name = &quot;XeniumR1&quot;)
Xen_R2 &lt;- importXenium(&quot;Xenium_R2/outs&quot;, sample_name = &quot;XeniumR2&quot;)</code></pre>
<p>Before moving on to the downstream analysis of the imaging-based
data, we can inspect both Xenium images. We use the
<strong>vrImages</strong> function to call and visualize reference
images of all VoltRon objects. Observe that the DAPI image of the second
Xenium replicate is not bright enough.</p>
<pre class="r watch-out"><code>vrImages(Xen_R1)
vrImages(Xen_R2)</code></pre>
<table>
<tbody>
<tr style="vertical-align: center">
<td style="width:50%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/xeniumr1.png" class="center">
</td>
<td style="width:50%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/xeniumr2.png" class="center">
</td>
</tr>
</tbody>
</table>
<p><br></p>
<p>We can adjust the brightness of the second Xenium replicate using the
<strong>modulateImage</strong> function where we can change the
brightness and` saturation of the reference image of this VoltRon
object. This functionality is optional for VoltRon objects and should be
used when images require further adjustments.</p>
<pre class="r watch-out"><code>Xen_R2 &lt;- modulateImage(Xen_R2, brightness = 800)
vrImages(Xen_R2)</code></pre>
<p><img width="40%" height="40%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/xeniumr2_new.png" class="center"></p>
<p><br></p>
<p>Once both VoltRon objects are created and images are well-tuned, we
can merge these two into a single VoltRon object.</p>
<pre class="r watch-out"><code>Xen_list &lt;- list(Xen_R1, Xen_R2)
Xen_data &lt;- merge(Xen_list[[1]], Xen_list[-1])</code></pre>
<pre><code>VoltRon Object 
XeniumR1: 
  Layers: Section1 
XeniumR2: 
  Layers: Section1 
Assays: Xenium(Main) </code></pre>
<p><br></p>
</div>
<div id="spatial-visualization" class="section level2">
<h2>Spatial Visualization</h2>
<p>With <strong>vrSpatialPlot</strong>, we can visualize subcellular
components as well as cells. We first visualize mRNAs of ACTA2, a marker
for smooth muscle cell actin, and TCF7, an early exhausted t cell
marker. We can interactively select a subset of the tissue section and
visualize the localization of these transcripts.</p>
<pre class="r watch-out"><code>Xen_R1_subsetinfo &lt;- subset(Xen_R1, interactive = TRUE)
Xen_R1_subset &lt;- Xen_R1_subsetinfo$subsets[[1]]
vrSpatialPlot(Xen_R1_subset, transcripts = c(&quot;ACTA2&quot;, &quot;TCF7&quot;), pt.size = 0.2)</code></pre>
<p><img width="50%" height="50%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_transcripts_visualize.png" class="center"></p>
<p>We can also visualize cell counts on spatially-aware data. The
behaviour of <strong>vrSpatialFeaturePlot</strong> (and most plotting
functions) depend on the number of assays associated with the assay type
(e.g. Xenium). Here, we have to assays, and we visualize two features,
hence the resulting plot would include four figures (two features for
both assays).</p>
<p>We can either use the original main image as a background, which is
the default behaviour of both vrSpatialFeaturePlot and vrSpatialPlot
functions, or select either “black” or “white” backgrounds
(e.g. <strong>background = “white”</strong>) as an additional arguement
in both functions.</p>
<pre class="r watch-out"><code>vrSpatialFeaturePlot(Xen_data, features = c(&quot;ACTA2&quot;, &quot;TCF7&quot;), alpha = 1, pt.size = 0.7)</code></pre>
<p><img width="90%" height="90%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_spatialfeature_xenium.png" class="center"></p>
</div>
<div id="processing-and-embedding" class="section level2">
<h2>Processing and Embedding</h2>
<p>Some number of cells in both Xenium images might have extremely low
molecule counts. Although cells are detected at these locations, the low
total counts of cells would make it challenging for phenotyping and
clustering these cells. Hence, we remove such cells from the VoltRon
objects.</p>
<pre class="r watch-out"><code>Xen_data &lt;- subset(Xen_data, Count &gt; 5)</code></pre>
<p>Later, we can move on to normalizing the counts to correct for count
depth of cells by dividing counts with total counts in each cell,
multiply with some constant (default: 10000), and followed by log
transformation of the counts.</p>
<pre class="r watch-out"><code>Xen_data &lt;- normalizeData(Xen_data)</code></pre>
<p>VoltRon is capable of reducing dimensionality of datasets using both
PCA and UMAP which we gonna use to build neighborhood graphs and
partition the data into cell types.</p>
<pre class="r watch-out"><code>Xen_data &lt;- getPCA(Xen_data, dims = 20)
Xen_data &lt;- getUMAP(Xen_data, dims = 1:20)</code></pre>
<p>We can also visualize the normalized expression of these features on
embedding spaces (e.g. UMAP) using
<strong>vrEmbeddingFeaturePlot</strong> function.</p>
<pre class="r watch-out"><code>vrEmbeddingFeaturePlot(Xen_data, features = c(&quot;LRRC15&quot;, &quot;TCF7&quot;), embedding = &quot;umap&quot;, pt.size = 0.4)</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_featureplot_xenium.png" class="center"></p>
</div>
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<p>Next, we build neighborhood graphs with the <strong>shared nearest
neighbor (SNN) graph</strong> of cells which are constructed from
dimensionally reduced gene expression profiles. The function
<strong>getProfileNeighbors</strong> also has an option of building
<strong>k-nearest neighbors (kNN)</strong> graphs. We can later conduct
a clustering of cells using the <strong>leiden’s method</strong> from
the igraph package, which is utilized with the
<strong>getClusters</strong> function.</p>
<pre class="r watch-out"><code>Xen_data &lt;- getProfileNeighbors(Xen_data, dims = 1:20, k = 10, method = &quot;SNN&quot;)
Xen_data &lt;- getClusters(Xen_data, resolution = 1.0, label = &quot;Clusters&quot;, graph = &quot;SNN&quot;)</code></pre>
<p>Now we can label each cell with the associated clustering index and
take a look at the clustering accuracy on the embedding space.</p>
<pre class="r watch-out"><code>vrEmbeddingPlot(Xen_data, group.by = &quot;Clusters&quot;, embedding = &quot;umap&quot;, pt.size = 0.4)</code></pre>
<p><img width="60%" height="60%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_embedplot_xenium.png" class="center"></p>
<p>We can also visualize these clusters on a spatial context.</p>
<pre class="r watch-out"><code>vrSpatialPlot(Xen_data, group.by = &quot;Clusters&quot;, pt.size = 0.4, background = &quot;black&quot;)</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_spatial_xenium.png" class="center"></p>
<p><br></p>
</div>
<div id="cell-neighborhood-analysis" class="section level2">
<h2>Cell Neighborhood Analysis</h2>
<p>We use the <strong>vrNeighbourhoodEnrichment</strong> function to
detect cell type pairs that co occur within each other neighborhoods.
First, we establish spatial neighborhood graphs that determine the
neighbors of each cell on tissue sections. <strong>Delaunay
tesselations</strong> or graphs are commonly used to determine neighbors
of spatial entities. The function <strong>getSpatialNeighbors</strong>
builds a delaunay graph of all assays of a certain type and detects
neighbors of cells in a VoltRon object.</p>
<pre class="r watch-out"><code>vr_merged_result &lt;- getSpatialNeighbors(vr_merged, method = &quot;delaunay&quot;)</code></pre>
<p>Once neighbors are founds, we can apply a permutation test compares
the number of cell type occurances with an expected number of these
occurances under multiple permutations of labels in the tissue (fixed
coordinates but cells are randomly labelled). A similar approach is used
to by several spatial analysis frameworks and packages (<a
href="https://www.nature.com/articles/nmeth.4391">Schapiro et. al
2017</a>, <a
href="https://www.nature.com/articles/s41592-021-01358-2">Palla et. al
2022</a>).</p>
<pre class="r watch-out"><code>neighborhood_results &lt;- vrNeighbourhoodEnrichment(vr_merged_result, group.by = &quot;Clusters&quot;)</code></pre>
<p>The neighbor analysis provides the results of both the association
and segregation tests between all cell type pairs across each layer and
assay. Here, the number of each cell in a pair in each section is
reported to assess the impact of the results of the test (i.e. low
number of abundance in one cell type may indicate low impact).</p>
<pre class="r watch-out"><code>head(neighborhood_results)</code></pre>
<div>
<pre><code style="font-size: 10px;">          from_value     to_value   p_assoc   p_segreg p_assoc_adj p_segreg_adj n_from n_to AssayID Assay    Layer         Sample
Assay1.1 CD163+ macs  CD163+ macs 0.0000000 1.00000000      0.0000   1.00000000     41   41  Assay1  MELC Section1 control_case_3
Assay1.2 CD163+ macs CD4+ T cells 0.9380000 0.03300000      0.9980   0.09762866     41   48  Assay1  MELC Section1 control_case_3
Assay1.3 CD163+ macs  CD8+ Tcells 0.8779011 0.04339051      0.9980   0.09762866     41   11  Assay1  MELC Section1 control_case_3
Assay1.4 CD163+ macs     NK cells 0.8190000 0.08700000      0.9980   0.15660000     41   15  Assay1  MELC Section1 control_case_3
Assay1.5 CD163+ macs   endothelia 0.1230000 0.85100000      0.5535   0.95737500     41  139  Assay1  MELC Section1 control_case_3
Assay1.6 CD163+ macs    epithelia 0.9320000 0.03600000      0.9980   0.09762866     41   39  Assay1  MELC Section1 control_case_3</code></pre>
</div>
<p><br></p>
</div>
</div>
<div id="melc-data-analysis" class="section level1">
<h1>MELC Data Analysis</h1>
<p>Other tham sequencing and imaging based spatial omics assays, VoltRon
provides support for imaging based proteomics assays as well.</p>
<p>In this next use case, we analyze segmented cells over microscopy
images collected from <strong>control</strong> and
<strong>COVID-19</strong> lung tissues of donors categorized based on
disease durations (<strong>control</strong>, <strong>acute</strong>,
<strong>chronic</strong> and <strong>prolonged</strong>). Each image is
associated with one of few field of views (FOVs) from a single tissue
section of a donor. See <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190732">GSE190732</a>
for more information. You can download the <strong>IFdata.csv</strong>
file and the folder with the <strong>DAPI</strong> images <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Cellanalysis/MELC/GSE190732.zip">here</a>.</p>
<p>We analyze cells characterized by <strong>multi-epitope ligand
cartography (MELC)</strong> with a panel of 44 parameters. We use the
already segmented cells on which expression of <strong>43 protein
features</strong> (excluding DAPI) were mapped to these cells.</p>
<p>We import the <strong>protein intensities</strong>,
<strong>metadata</strong> and <strong>coordinates</strong> associated
with segmented cells across FOVs of samples.</p>
<pre class="r watch-out"><code>library(VoltRon)
IFdata &lt;- read.csv(&quot;IFdata.csv&quot;)
data &lt;- IFdata[,c(2:43)]
metadata &lt;- IFdata[,c(&quot;disease_state&quot;, &quot;object_id&quot;, &quot;cluster&quot;, &quot;Clusters&quot;,
                      &quot;SourceID&quot;, &quot;Sample&quot;, &quot;FOV&quot;, &quot;Section&quot;)]
coordinates &lt;- as.matrix(IFdata[,c(&quot;posX&quot;,&quot;posY&quot;)], rownames.force = TRUE)</code></pre>
<p><br></p>
<div id="importing-melc-data" class="section level2">
<h2>Importing MELC data</h2>
<p>First, we normalize the data using the hyperbolic arcsine function
with a scale argument of 0.2.</p>
<pre class="r watch-out"><code>data &lt;- asinh(data/0.2)</code></pre>
<p>Before analyzing MELC assays across FOVs, we should <strong>build a
VoltRon object</strong> for each individual FOV/Section by using the
<strong>formVoltron</strong> function. We then merge these sections to
respective tissue blocks by defining their samples of origins. We can
also define <strong>assay names</strong>, <strong>assay types</strong>
and <strong>sample (i.e. block) names</strong> of these objects.</p>
<pre class="r watch-out"><code>library(dplyr)
library(magick)
vr_list &lt;- list()
sample_metadata &lt;- metadata %&gt;% select(Sample, FOV, Section) %&gt;% distinct()
for(i in 1:nrow(sample_metadata)){
  vrassay &lt;- sample_metadata[i,]
  cells &lt;- rownames(metadata)[metadata$Section == vrassay$Section]
  image &lt;- image_read(paste0(&quot;DAPI/&quot;, vrassay$Sample, &quot;/DAPI_&quot;, vrassay$FOV, &quot;.tif&quot;))
  vr_list[[vrassay$Section]] &lt;- formVoltRon(data = t(data[cells,]), 
                                            metadata = metadata[cells,],
                                            image = image, 
                                            coords = coordinates[cells,],
                                            main.assay = &quot;MELC&quot;, 
                                            assay.type = &quot;cell&quot;,
                                            sample_name = vrassay$Section)
}</code></pre>
<p>Before moving forward with merging sections, we should <strong>flip
coordinates</strong> of cells and perhaps also then
<strong>resize</strong> these images. The main reason for this
coordinate flipping is that the y-axis of most digital images are of the
opposite direction to the commonly used coordinate spaces. Hence,
VoltRon offers the option to flip coordinates of spatial points with
respect to the image using <strong>flipCoordinates</strong>
function.</p>
<pre class="r watch-out"><code>for(i in 1:nrow(sample_metadata)){
  vrassay &lt;- sample_metadata[i,]
  vr_list[[vrassay$Section]] &lt;- flipCoordinates(vr_list[[vrassay$Section]])
  vr_list[[vrassay$Section]] &lt;- resizeImage(vr_list[[vrassay$Section]], size = 600)
}</code></pre>
<p>Finally, we merge these assays into one VoltRon object. The
<strong>samples</strong> arguement in the merge function determines
which assays are layers of a single tissue sample/block.</p>
<pre class="r watch-out"><code>vr_merged &lt;- merge(vr_list[[1]], vr_list[-1], samples = sample_metadata$Sample)
vr_merged </code></pre>
<pre><code>VoltRon Object 
control_case_3: 
  Layers: Section1 Section2 
control_case_2: 
  Layers: Section1 Section2 
control_case_1: 
  Layers: Section1 Section2 Section3 
acute_case_3: 
  Layers: Section1 Section2 
acute_case_1: 
  Layers: Section1 Section2 
... 
There are 13 samples in total 
Assays: MELC(Main) </code></pre>
<p><br></p>
<p>The first control case have three fields of views (FOVs). By
subsetting on the sample of the first control case, we can visualize
only these three sections, and visualize multiple protein markers.</p>
<pre class="r watch-out"><code>vr_subset &lt;- subset(vr_merged, samples = &quot;control_case_1&quot;)
vrSpatialFeaturePlot(vr_subset, features = c(&quot;CollagenIV&quot;, &quot;CD11b&quot;), alpha = 1, pt.size = 0.7)</code></pre>
<p><img width="90%" height="90%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_spatialfeature.png" class="center"></p>
<p><br></p>
</div>
<div id="dimensionality-reduction" class="section level2">
<h2>Dimensionality Reduction</h2>
<p>We can utilize dimensional reduction of the available protein markers
using the getPCA and getUMAP functions, but now with lower numbers of
principal components which are enough to capture information marker list
of 44 proteins.</p>
<pre class="r watch-out"><code>vr_merged &lt;- getPCA(vr_merged, dims = 10)
vr_merged &lt;- getUMAP(vr_merged, dims = 1:10)
vrEmbeddingFeaturePlot(vr_merged, features = c(&quot;CollagenIV&quot;, &quot;CD11b&quot;), embedding = &quot;umap&quot;)</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_embedding.png" class="center"></p>
<p><br></p>
</div>
<div id="clustering-1" class="section level2">
<h2>Clustering</h2>
<p>Now we can visualize the clusters across these sections and perhaps
also check for clusters that may reside in only specific disease
conditions.</p>
<pre class="r watch-out"><code># SNN graph and clusters
vr_merged &lt;- getProfileNeighbors(vr_merged, dims = 1:10, k = 10, method = &quot;kNN&quot;)
vr_merged &lt;- getClusters(vr_merged, resolution = 0.8, label = &quot;MELC_Clusters&quot;)

# visualize conditions and clusters
library(patchwork)
vr_merged$Condition &lt;- gsub(&quot;_[0-9]$&quot;, &quot;&quot;, vr_merged$Sample)
g1 &lt;- vrEmbeddingPlot(vr_merged, group.by = c(&quot;Condition&quot;), embedding = &quot;umap&quot;)
g2 &lt;- vrEmbeddingPlot(vr_merged, group.by = c(&quot;MELC_Clusters&quot;), embedding = &quot;umap&quot;, 
                      label = TRUE)
g1 | g2</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_embeddingclusters.png" class="center"></p>
<p><br></p>
</div>
<div id="visualization-of-markers" class="section level2">
<h2>Visualization of Markers</h2>
<p>VoltRon provides both violin plots (<strong>vrViolinPlot</strong>)
and heatmaps (<strong>vrHeatmapPlot</strong>) to further investigate the
enrichment of markers across newly clustered datasets.
<strong>Note:</strong> the vrHeatmapPlot function would require you to
have the <strong>ComplexHeatmap</strong> package in your namespace.</p>
<pre class="r watch-out"><code>library(ComplexHeatmap)
vrHeatmapPlot(vr_merged, features = vrFeatures(vr_merged), 
              group.by = &quot;MELC_Clusters&quot;, show_row_names = TRUE)</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_heatmapclusters.png" class="center"></p>
<p><br></p>
<pre class="r watch-out"><code>vrViolinPlot(vr_merged, features = c(&quot;CD3&quot;, &quot;SMA&quot;, &quot;Pancytokeratin&quot;, &quot;CCR2&quot;), 
             group.by = &quot;MELC_Clusters&quot;, ncol = 2)</code></pre>
<p><img width="100%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_violinclusters.png" class="center"></p>
<p><br></p>
</div>
<div id="neighborhood-analysis" class="section level2">
<h2>Neighborhood Analysis</h2>
<pre class="r watch-out"><code>vr_merged_result &lt;- getSpatialNeighbors(vr_merged, method = &quot;delaunay&quot;)
neighborhood_results &lt;- vrNeighbourhoodEnrichment(vr_merged_result, group.by = &quot;Clusters&quot;)</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
