---
title: "Tutorials"
---

<style>
.title{
  display: none;
}
body {
  text-align: justify
}
table, th, td, tr{
  empty-cells:hide;
  border-spacing: 25px;
  border-collapse: separate;
}
td{
  filter:drop-shadow(0 0 10px rgba(0,0,0,0.3));
  padding: 10px;
  background:white;
  text-align:center;
  width:33%;
  vertical-align: text-top;
}
td:hover, td:active{
  filter:drop-shadow(0 0 10px rgba(0,0,0,0.3));
  transform: scale(1.02);
  cursor: pointer;
}
.main-container {
  max-width: 1200px;
  margin-left: auto;
  margin-right: auto;
}
p.tutorial {
  text-decoration: none!important;
  font-size: 1.5em;
  color: #23803A;
  margin: 3%
}
</style>

```{r setup, include=FALSE}
# use rmarkdown::render_site(envir = knitr::knit_global())
knitr::opts_chunk$set(echo = TRUE)
```

<br>
<!-- <br> -->

<!-- ## Spatial Data Analysis with VoltRon -->

<br>

## Spatially Aware Analysis

**VoltRon** will incorporate multiple data integration modules to achieve data transfer across a diverse set of modalities. We utilize the <a href="https://opencv.org/">OpenCV</a> library which fully embedded into VoltRon using <a href="https://www.rcpp.org/">Rcpp</a>, hence VoltRon can align reference images of spatial datasets using computer vision and image registration. Users can automaticaly or manually align a list of images (**H&E, microscopy or DAPI images**) using a small shiny app incorporated within our analysis workflow. 

In addition, VoltRon integrates with spot deconvolution methods such as RCTD to infer spot transcriptomics assays such as Visium and DBIT-Seq. We estimate the cell type abundances from a reference single cell data (Seurat, SingleCellExperiment etc.) with already annotated cell types. Then we generate additional assays associated with layers of the VoltRon object, and implement **clustering on normalized cell type abundance data** to discover niches in the dataset. 

<table>
<tbody>
  <tr>
   <td onclick="location.href='registration.html';"> 
      <p class ="tutorial"> Spatial Data Alignment </p>
      <img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/registration.png" class="center">
      <p style = "margin: 3%"> Automated and manual alignment of spatial data assays</p>
   </td>
   <td onclick="location.href='multiomic.html';">
      <p class ="tutorial"> Multi-omic Integration </p>
      <img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic.png" class="center">
      <p style = "margin: 3%"> Integrating data modalities within or across tissue sections </p>
   </td>
   <td onclick="location.href='deconvolution.html';"> 
      <p class ="tutorial"> Niche Clustering </p>
      <img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/deconvolution.png" class="center">
      <p style = "margin: 5%"> Clustering based on ROI/spot deconvolution and Spatial Neighborhood </p>
   </td>
  </tr>
</tbody>
</table>

<br>
<br>

## Downstream Analysis

**VoltRon** is also capable of end-to-end analysis of diverse set of spatial data types such as **ROIs** (regions of interest), **spots**, **single cells**,  **molecules** and even **images**. VoltRon incorporates built-in functions to import readouts of these distinct spatial omic technologies where a similar set of functions can be used to analyze each individual data type.

<table>
<tbody>
  <tr>
   <td onclick="location.href='roianalysis.html';"> 
      <p class ="tutorial"> Region of Interests (ROIs) </p>
      <img width="85%" height="85%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/GeoMx.png" class="center">
      <p style = "margin: 3%"> Quality control, analysis and visualization of ROI segments </p>
   </td>
   <td onclick="location.href='spotanalysis.html';"> 
      <p class ="tutorial"> Cells/Spots </p>
      <img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/spotanalysis.png" class="center">
      <p style = "margin: 3%"> Quality control, analysis and visualization of Cell/Spot datasets </p>
   </td>
      <td onclick="location.href='moleculeanalysis.html';"> 
      <p class ="tutorial"> Molecules </p>
      <br>
      <img width="85%" height="85%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_visualize.png" class="center">
      <br>
      <p style = "margin: 3%"> Analysis and visualization of Molecule datasets <br> <strong> (Under Development) </strong> </p>
   </td>
  </tr>
  <tr>
     <td onclick="location.href='pixelanalysis.html';"> 
      <p class ="tutorial"> Pixels (Image Only) </p>
      <img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/tissue_lowres_image_grid.png" class="center">
      <p style = "margin: 3%"> Analysis and visualization of Image datasets <br> <strong> (Under Development) </strong> </p>
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

<br>
<br>

##  Other Utilities

Here, we provide a group of tutorials to use additional features of the VoltRon objects. Here we describe a collection of features that VoltRon package utilizes such as interactive analysis and importing spatially aware data from diverse spatial omic technologies. VoltRon will be able to convert its objects to a diverse set of objects/datatypes commonly incorporated spatial data analysis (Seurat, SpatialExperiment, Giotto, anndata, SpatialData etc.) 

<table>
<tbody>
  <tr>
   <td onclick="location.href='interactive.html';"> 
      <p class ="tutorial"> Interactive Utilities </p>
      <img width="85%" height="85%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/interactiveannotation.png" class="center">
      <p style = "margin: 3%"> Interactive annotation and visualization  <br>  <strong> (Under Development) </strong> </p>
   </td>
   <td onclick="location.href='voltronobjects.html';"> 
      <p class ="tutorial"> Working with VoltRon Objects </p>
      <br>
      <img width="85%" height="85%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/voltronobjects.png" class="center">
      <br>
      <br>
      <p style = "margin: 3%"> Manipulating and configuring VoltRon objects </p>
   </td>
   <td onclick="location.href='importingdata.html';"> 
      <p class ="tutorial"> Importing Spatial Data </p>
      <img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/technologies.png" class="center">
      <p style = "margin: 3%"> Importing readouts of spatial technologies <br> </p>
   </td>
  </tr>
  <tr>
    <td onclick="location.href='conversion.html';"> 
      <p class ="tutorial"> Converting VoltRon Objects </p>
      <br>
      <img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/conversion.png" class="center">
      <br>
      <br>
      <p style = "margin: 3%"> Converting VoltRon objects into Seurat,de Squidpy (anndata) etc. </strong></p>
   </td>
   <td onclick="location.href='ondisk.html';"> 
      <p class ="tutorial"> OnDisk Analysis Utilities </p>
      <br>
      <img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/ondisk.png" class="center">
      <br>
      <br>
      <p style = "margin: 3%"> Efficient access to large VoltRon objects </strong></p>
   </td>
  </tr>
</tbody>
</table>

<br>
